### ISSUE TITLE
Correct way to support multiple checkboxes with same name

### ISSUE THREAD

**Issue by piranm** [react-hook-form/react-hook-form#476]
**Describe the question?**

What's the 'correct' way to support multiple checkboxes, where the user can pick one or more values in the same group? E.g., pick zero or more colors.

**To Reproduce**

In pure HTML I'd write inputs with the same name:

```html
    <input type="checkbox" name="colors" value="red">
    <input type="checkbox" name="colors" value="green">
    <input type="checkbox" name="colors" value="blue">
```

In react-hook-forms, how would I get the list of checked colors back as a list?

The obvious solution is to copy the above with the same name, but that doesn't produce the correct value. However, this just produces the value of the first checkbox (checked or not).

```JSX
// Doesn't work
rainbow.map(
    (c,i) => <label key={c}><input type="checkbox" value={c} name="sameName" ref={register} />{c}</label>
)
```

A working solution is to trick react-hook-forms into make the name an array, e.g., `colors.0`, which gives me an array of the checked checkboxes, but this feels like a hack.

```JSX
// Produces the correct outcome
rainbow.map(
    (c,i) => <label key={c}><input type="checkbox" value={c} name={"withIndex."+i} ref={register} />{c}</label>
)
````

**Codesandbox link**
https://codesandbox.io/s/react-hook-form-multiple-checkboxes-u95u2

Left-hand-side shows using the HTML style same name; clicking Submit just returns the value of the first checkbox.

Right-hand-side shows using a fake index, which appears to work correctly.

**Additional context**
(in my real app, the list of values is very dynamic)

**Comment by bluebill1049** [comment]
On my phone, couldn’t take a look ur coddsandbox but you just need to register them all.

**Comment by bluebill1049** [comment]
oh looks like that's a bug... we should support multiple checkbox

**Comment by bluebill1049** [comment]
hmm looking closely, I would actually recommend your second approach. It's tricky to determine whether a checkbox has single or multiple values. It would probably introduce quite bit complexity to support it.

**Comment by piranm** [comment]
Thanks @bluebill1049 — I've updated my question to make it clearer as well.

**Comment by bluebill1049** [comment]
thanks, @piranm let's stick with the input name with index.

**Comment by mlajtos** [comment]
@piranm Beware, this behavior broke in 3.18.7. In 3.18.6 you get empty array when nothing is selected, which is what you expect. In 3.18.7 and later, array contains `false` value for every non-selected checkbox.

**Comment by bluebill1049** [comment]
I am going to close this issue as I suggested above, let's stick with the name with index.

**Comment by bluebill1049** [comment]
Tmr's release will fix the array checkbox issue. you can do the following

```
   <input type="checkbox" name="colors" value="red">
    <input type="checkbox" name="colors" value="green">
    <input type="checkbox" name="colors" value="blue">```

**Comment by mlajtos** [comment]
That is great to hear, thank you. :) One question: what shape will the colors array have?

**Comment by bluebill1049** [comment]
same data shape. @mlajtos

**Comment by piranm** [comment]
My worry about dropping the array subscript (as proposed by @bluebill1049) is that I don't always know how many colors I want to render, and some may not be available (no checkbox rendered for that color).

So in the case of multiple colour checkboxes rendered, I would get an array of values, possibly empty.
In the case of a single colour checkbox rendered, I would get a single value, or false/undefined?

So we get different behaviour if we only render one checkbox, or the list of colours is a single value.

Perhaps better to have something in the `register` function to say that this is part of an array, or leave the subscript?

**Comment by bluebill1049** [comment]
@piranm we are not dropping anything, it's not a breaking change. We only add support for checkbox support with the same name.

**Comment by piranm** [comment]
So, to summarize from description, recent commit, and behaviour seen in 3.27.0:

* without an index, same name will return:
  * with one checkbox
    * if has value
      * if ticked, then the value
      * if not ticked, then false
    * if does not have value
      * if ticked, then true
      * if not ticked, then false
  * with two checkbox (must have values, I presume)
    * if one ticked, then array with checkbox value
    * if two ticked, then array with both checkbox values
    * if none ticked, then empty array
* with an index on the name, will return:
  * with one checkbox
    * if ticked, a list with the value, e.g, ["red"]
    * if not ticked, then a list with false, e.g., [false]
  * with two checkboxes
    * if one ticked, a list with the values or false, e.g, ["red", false]
    * if two ticked, a list with the value, e.g, ["red", "green"]
    * if none ticked, then a list with false, e.g., [false, false]
  * with checkboxes with gaps a missing index 1
    * those gaps will be undefined, e.g. ["red", undefined, false]

I'm happy with that, probably sticking to using the index version.

**Comment by bluebill1049** [comment]
The above logic looks correct 👍 

Yea I think for your use case, it's better to stick with index version. (there is no point of changing too)

**Comment by chrisrobbins** [comment]
@bluebill1049 I have a special case where the index won't work. My data structure:

`[ section: string, data: [] ]`

Here I map through the sections, then map through the data and create a checkbox for each object in `data`. This is problematic because the objects in each section will repeat indices. I found a workaround by using an `ID` from the object, but it delivers over a hundred empty `{}` in the submit data. 
My question is, is there a better way to do what I am trying to do, or is there some reason and way to avoid all of the empty `{}`?

**Comment by bluebill1049** [comment]
can you provide a simple codesandbox? @chrisrobbins

**Comment by chrisrobbins** [comment]
Sure!
https://codesandbox.io/s/react-hook-form-checkbox-arrays-80z3q

**Comment by bluebill1049** [comment]
I see your problem with empty array fields, unfortunately, that's how we organize data with array-like syntax, you probably have to use `filter(Boolean)`

**Comment by pom421** [comment]
I have a similar problem. I have a bunch of checkbox, and bad luck, a group of checkbox has only one value. So this group is considered like a boolean, whereas the others are considered array of string. 

I have seen the answer with [] syntax but in this case, the potential false values are not relevant and I find that is awkward to have it.

I have made a workaround with a visually hidden checkbox alongside with my single checkbox in a group, and this way RHF interpret the group like an array, but it's not great.

**Comment by stage88** [comment]
Hi everyone, here is something similar that has worked for me, sharing in case anyone else encountered this problem. 

https://codesandbox.io/s/crazy-framework-647uk

**Comment by bluebill1049** [comment]
> Hi everyone, here is something similar that has worked for me, sharing in case anyone else encountered this problem.
> 
> https://codesandbox.io/s/crazy-framework-647uk

Thanks for sharing ❤️

**Comment by simonho1025** [comment]
> Hi everyone, here is something similar that has worked for me, sharing in case anyone else encountered this problem.
> 
> https://codesandbox.io/s/crazy-framework-647uk
@stage88 @bluebill1049 Anyone know how to do validation using Yup on this example? Such as must select at least one or two each group.

**Comment by joshunger** [comment]
If you're looking for `setValue` usage this works correctly -

```javascript
const color = ...
setValue(
  'colors',
  getValues('colors').filter(value => value !== color),
)
```

**Comment by graemetavour** [comment]
I ran into this issue and eventually found the discussion here (very helpful, thanks!). Would it be possible to get some clearer documentation at https://react-hook-form.com/api/useform/register? 

That was the first place I went and while it has a little documentation about how to work with an array of fields, it's not totally clear especially since the correct way to handle this generates an array with `false`s mixed in (I wouldn't have expected that to be the right way to handle this.

Basically, it would be great if everything in this comment https://github.com/react-hook-form/react-hook-form/issues/476#issuecomment-553849830 were in the docs, along with a link to a codesandbox example.

**Comment by penx** [comment]
@bluebill1049 as above, I _really_ think 'how to handle checkboxes' should be in the getting started or FAQ docs...

Please could you reopen this?

**Comment by noahdecoco** [comment]
Totally agree, there's no mention of how to handle a group of checkboxes the "right" way anywhere in the documentation. I think it would be immensely useful to have an example in the getting started section.

**Comment by jjenzz** [comment]
In case it helps anyone, this appears to be the only way i could get an array of checkbox **values** https://codesandbox.io/s/naughty-matsumoto-iozfj

**Comment by GoldenAgeMike** [comment]
> In case it helps anyone, this appears to be the only way I could get an array of checkbox **values** when using `Controller`: https://codesandbox.io/s/naughty-matsumoto-iozfj

That helped alot with the time I have . This works perfectly.

**Comment by bluebill1049** [comment]
thanks, @jjenzz for sharing the solution: 

here are my 2cents on how to use checkboxes with uncontrolled and controlled components:

https://codesandbox.io/s/dark-voice-99ld4?file=/src/App.js

```tsx
import * as React from "react";
import { useController, useForm } from "react-hook-form";
import Headers from "./Header";
import "./styles.css";

let renderCount = 0;

const Checkboxes = ({ options, control, name }) => {
  const { field } = useController({
    control,
    name
  });
  const [value, setValue] = React.useState(field.value || []);

  return (
    <>
      {options.map((option, index) => (
        <input
          onChange={(e) => {
            const valueCopy = [...value];

            // update checkbox value
            valueCopy[index] = e.target.checked ? e.target.value : null;

            // send data to react hook form
            field.onChange(valueCopy);

            // update local state
            setValue(valueCopy);
          }}
          key={option}
          type="checkbox"
          value={option}
        />
      ))}
    </>
  );
};

export default function App() {
  const { register, handleSubmit, control } = useForm({
    defaultValues: {
      controlled: [],
      uncontrolled: []
    }
  });
  const onSubmit = (data) => console.log(data);
  renderCount++;

  return (
    <div>
      <Headers
        renderCount={renderCount}
        description="Performant, flexible and extensible forms with easy-to-use validation."
      />

      <form onSubmit={handleSubmit(onSubmit)}>
        <section>
          <h2>uncontrolled</h2>
          <input {...register("uncontrolled")} type="checkbox" value="A" />
          <input {...register("uncontrolled")} type="checkbox" value="B" />
          <input {...register("uncontrolled")} type="checkbox" value="C" />
        </section>

        <section>
          <h2>controlled</h2>
          <Checkboxes
            options={["a", "b", "c"]}
            control={control}
            name="controlled"
          />
        </section>
        <input type="submit" />
      </form>
    </div>
  );
}
```

both examples will be included in the doc within the code section under `register` and `useController`.

**Comment by enheit** [comment]
I created a [codesandbox example](https://codesandbox.io/s/gallant-colden-o05ez?file=/src/App.tsx) that shows how to use multiple checkboxes with react-hook-form + react material ui. Hope it will be helpful for somebody.

**Comment by bluebill1049** [comment]
> I created a [codesandbox example](https://codesandbox.io/s/gallant-colden-o05ez?file=/src/App.tsx) that shows how to use multiple checkboxes with react-hook-form + react material ui.
> 
> Hope this will be helpful for someone

awesome thanks for sharing. Here is an native one: https://codesandbox.io/s/usecontroller-checkboxes-99ld4

in the doc as well: https://react-hook-form.com/api/usecontroller

**Comment by herialesha** [comment]
> thanks, @jjenzz for sharing the solution:
> 
> here are my 2cents on how to use checkboxes with uncontrolled and controlled components:
> 
> https://codesandbox.io/s/dark-voice-99ld4?file=/src/App.js
> 
> ```tsx
> import * as React from "react";
> import { useController, useForm } from "react-hook-form";
> import Headers from "./Header";
> import "./styles.css";
> 
> let renderCount = 0;
> 
> const Checkboxes = ({ options, control, name }) => {
>   const { field } = useController({
>     control,
>     name
>   });
>   const [value, setValue] = React.useState(field.value || []);
> 
>   return (
>     <>
>       {options.map((option, index) => (
>         <input
>           onChange={(e) => {
>             const valueCopy = [...value];
> 
>             // update checkbox value
>             valueCopy[index] = e.target.checked ? e.target.value : null;
> 
>             // send data to react hook form
>             field.onChange(valueCopy);
> 
>             // update local state
>             setValue(valueCopy);
>           }}
>           key={option}
>           type="checkbox"
>           value={option}
>         />
>       ))}
>     </>
>   );
> };
> 
> export default function App() {
>   const { register, handleSubmit, control } = useForm({
>     defaultValues: {
>       controlled: [],
>       uncontrolled: []
>     }
>   });
>   const onSubmit = (data) => console.log(data);
>   renderCount++;
> 
>   return (
>     <div>
>       <Headers
>         renderCount={renderCount}
>         description="Performant, flexible and extensible forms with easy-to-use validation."
>       />
> 
>       <form onSubmit={handleSubmit(onSubmit)}>
>         <section>
>           <h2>uncontrolled</h2>
>           <input {...register("uncontrolled")} type="checkbox" value="A" />
>           <input {...register("uncontrolled")} type="checkbox" value="B" />
>           <input {...register("uncontrolled")} type="checkbox" value="C" />
>         </section>
> 
>         <section>
>           <h2>controlled</h2>
>           <Checkboxes
>             options={["a", "b", "c"]}
>             control={control}
>             name="controlled"
>           />
>         </section>
>         <input type="submit" />
>       </form>
>     </div>
>   );
> }
> ```
> 
> both examples will be included in the doc within the code section under `register` and `useController`.

how to display error when on submit evidently not checked?

**Comment by cereallarceny** [comment]
@bluebill1049 Any idea about how to do this with `minLength` validation? I've been trying to figure out how to add a validation error if the user selects any less than 2 of the 5 available checkboxes. Something like that. Here's what I've been doing (integrated with ChakraUI), but with no luck so far:

```jsx
const CheckboxesInput = ({
  form: { control },
  rules,
  name,
  data: { options, key, value, description, orientation = 'vertical' },
}) => (
  <Controller
    name={name}
    rules={rules}
    control={control}
    render={({ field: { ref, onBlur, ...rest } }) => (
      <CheckboxGroup {...rest}>
        <Stack
          spacing={SPACING}
          direction={[
            'column',
            null,
            orientation === 'horizontal' ? 'row' : 'column',
          ]}
        >
          {options.map((v) => {
            const isString = typeof v === 'string';
            const fk = isString ? v : v[key];
            const fv = isString ? v : v[value];

            return (
              <Checkbox key={fk} m={0} value={fk} onBlur={onBlur}>
                {fv}
                {description && (
                  <Text fontSize="sm" color="gray.500">
                    {v[description]}
                  </Text>
                )}
              </Checkbox>
            );
          })}
        </Stack>
      </CheckboxGroup>
    )}
  />
);
```

**Comment by lulumetro** [comment]
> thanks, @jjenzz for sharing the solution:
> 
> here are my 2cents on how to use checkboxes with uncontrolled and controlled components:
> 
> https://codesandbox.io/s/dark-voice-99ld4?file=/src/App.js
> 
> ```tsx
> import * as React from "react";
> import { useController, useForm } from "react-hook-form";
> import Headers from "./Header";
> import "./styles.css";
> 
> let renderCount = 0;
> 
> const Checkboxes = ({ options, control, name }) => {
>   const { field } = useController({
>     control,
>     name
>   });
>   const [value, setValue] = React.useState(field.value || []);
> 
>   return (
>     <>
>       {options.map((option, index) => (
>         <input
>           onChange={(e) => {
>             const valueCopy = [...value];
> 
>             // update checkbox value
>             valueCopy[index] = e.target.checked ? e.target.value : null;
> 
>             // send data to react hook form
>             field.onChange(valueCopy);
> 
>             // update local state
>             setValue(valueCopy);
>           }}
>           key={option}
>           type="checkbox"
>           value={option}
>         />
>       ))}
>     </>
>   );
> };
> 
> export default function App() {
>   const { register, handleSubmit, control } = useForm({
>     defaultValues: {
>       controlled: [],
>       uncontrolled: []
>     }
>   });
>   const onSubmit = (data) => console.log(data);
>   renderCount++;
> 
>   return (
>     <div>
>       <Headers
>         renderCount={renderCount}
>         description="Performant, flexible and extensible forms with easy-to-use validation."
>       />
> 
>       <form onSubmit={handleSubmit(onSubmit)}>
>         <section>
>           <h2>uncontrolled</h2>
>           <input {...register("uncontrolled")} type="checkbox" value="A" />
>           <input {...register("uncontrolled")} type="checkbox" value="B" />
>           <input {...register("uncontrolled")} type="checkbox" value="C" />
>         </section>
> 
>         <section>
>           <h2>controlled</h2>
>           <Checkboxes
>             options={["a", "b", "c"]}
>             control={control}
>             name="controlled"
>           />
>         </section>
>         <input type="submit" />
>       </form>
>     </div>
>   );
> }
> ```
> 
> both examples will be included in the doc within the code section under `register` and `useController`.

@bluebill1049 I followed your example to implement a wrapper component for a group of checkboxes, but there is an issue with default values checked state being lost after cheking other values. I've forked your CSB removing the uncontrolled section and adding more controlled checkboxes as well as changing the default values. You can see how some of the default values are unchecked after checking other initially unchecked checkboxes. I'm having a hard time trying to get all this stuff working, so any help would be appreciated

[https://codesandbox.io/s/usecontroller-checkboxes-forked-lny46?file=/src/App.js](url)

**Comment by cereallarceny** [comment]
@bluebill1049 I modified the example present on the Code Examples section of the site: https://codesandbox.io/s/checkbox-min-checked-forked-0ljrm?file=/src/App.js

I suppose my main question is: **why can't we use the minLength rule to do this? Why does this only work with the validate function?**

**Comment by jorgen-kaxig** [comment]
> I created a [codesandbox example](https://codesandbox.io/s/gallant-colden-o05ez?file=/src/App.tsx) that shows how to use multiple checkboxes with react-hook-form + react material ui.
> 
> Hope this will be helpful for someone

Thanks a lot @enheit 
All the other examples I've seen, loops the ther "options array" around the entire Controller
But in your example you only have ONE SINGLE Controller and then loops checkboxes inside its "render" property.
That's just brilliant!!!

**Comment by nubpro** [comment]
> I created a [codesandbox example](https://codesandbox.io/s/gallant-colden-o05ez?file=/src/App.tsx) that shows how to use multiple checkboxes with react-hook-form + react material ui.
> 
> Hope this will be helpful for someone

Thanks for sharing, you're a legend! This saved me 2 days of headache
